[Device-capability] [Create and update document][1452622] (#2656)
MicrosoftDocs/msteams-docs: msteams-platform/concepts/device-capabilities/mobile-camera-image-permissions.md

---
title: Integrate media capabilities 
description: How to use Teams Javascript client SDK to enable media capabilities
keywords: camera image microphone capabilities native device permissions media
ms.topic: conceptual
ms.author: lajanuar
---

# Integrate media capabilities 

> This document gives a walkthrough on integrating  native device capabilities with Teams platform.

Reviewer: ```suggestion
This document gives a walkthrough on integrating media capabilities like camera, mic, speaker with your Teams app.
You can use the  [Microsoft Teams JavaScript client SDK](/javascript/api/overview/msteams-client?view=msteams-client-js-latest&preserve-view=true), to easily integrate these capabilities within your Microsoft Teams mobile app. The SDK provides the tools necessary for your app to access a user’s [device permissions](native-device-permissions.md?tabs=desktop#device-permissions) and build a richer experience.
```

> The process with which you combine the native device capabilities, such as **camera**, and **microphone** along with Teams platform is called integration of media capabilities. 

Reviewer: No need to explain what integration means as it's well understood in dev community.
```suggestion
```

## Integrate media capabilities - Advantage

The main advantage of integrating device capabilities with Teams platform is to enhance and enrich Teams collaborative experience.

> After accessing device permissions, integrate media capabilities by updating the app manifest file and calling the media capability APIs. 

Reviewer: ```suggestion
Integrate media capabilities by updating the app manifest file and calling the media capability APIs as documented below. 
```

> Have a good understanding on [code snippets](#code-snippets) for calling the respective APIs which allow you to use native media capabilities.

Reviewer: ```suggestion
You may want to refer [code snippets](#code-snippets) to build good understanding of how to make calls to different media APIs. It is also important to familiarize yourself with the [API response errors](#error-handling) so make sure you handle them appropriately in your Teams app.
```

> Before you proceed further, it is important to familiarize yourself with the [API response errors](#error-handling) that are generated while calling the APIs and the methods to handle them.

Reviewer: ```suggestion
```

---
title: Integrate media capabilities 
description: How to use Teams JavaScript client SDK to enable media capabilities
keywords: camera image microphone capabilities native device permissions media
ms.topic: conceptual
ms.author: lajanuar
---

# Integrate media capabilities 

This document guides you on how to integrate media capabilities. This integration combines the native device capabilities, such as the **camera** and **microphone** with the Teams platform.  

You can use [Microsoft Teams JavaScript client SDK](/javascript/api/overview/msteams-client?view=msteams-client-js-latest&preserve-view=true), that provides the tools necessary for your app to access a user’s [device permissions](native-device-permissions.md). Use suitable  **media capability APIs** to integrate the native device capabilities, such as the **camera** and **microphone** with the Teams platform within your Microsoft Teams mobile app, and build a richer experience. 

## Advantage of integrating media capabilities

> The main advantage of integrating device capabilities with the Teams platform is to enhance and enrich its innovative experience.

Reviewer: ```suggestion
The main advantage of integrating device capabilities in your Teams apps is it leverages native Teams controls to provide a rich and immersive experience to your end users.
```

> After accessing device permissions, integrate media capabilities by updating the app manifest file and calling the media capability APIs. 

Reviewer: ```suggestion
To integrate media capabilities you must update the app manifest file and call the media capability APIs. 
```

---
title: Integrate media capabilities 
description: How to use Teams JavaScript client SDK to enable media capabilities
keywords: camera image microphone capabilities native device permissions media
ms.topic: conceptual
ms.author: lajanuar
---

# Integrate media capabilities 

The process with which you combine the native device capabilities, such as the **camera** and **microphone** with the Teams platform is called the integration of media capabilities. 

> This document gives a walkthrough on integrating media capabilities like camera, microphone, and speaker with your Teams app.

Reviewer 2: Please update line 11 and 13. Seems wordy.

You can use the [Microsoft Teams JavaScript client SDK](/javascript/api/overview/msteams-client?view=msteams-client-js-latest&preserve-view=true), to integrate these capabilities easily within your Microsoft Teams mobile app. The SDK provides the tools necessary for your app to access a user’s [device permissions](native-device-permissions.md) and build a richer experience.

> ## Integrate media capabilities - Advantages

Reviewer 2: There is only one advantage as per the content under this heading. Please see if you want to modify the heading?

The main advantage of integrating device capabilities with the Teams platform is to enhance and enrich its innovative experience.
After accessing device permissions, integrate media capabilities by updating the app manifest file and calling the media capability APIs. 

For effective integration, you must have a good understanding of [code snippets](#code-snippets) for calling the respective APIs, which allow you to use native media capabilities.

It is important to familiarize yourself with the [API response errors](#error-handling) to handle the errors  in your Teams app.

> [!NOTE] 

> > Currently, the Microsoft Teams support for media capabilities is only available for mobile clients.

Reviewer 2: Remove article "the"

---
title: Media capabilities in Teams
description: How to use Teams Javascript client SDK to enable media capabilities
keywords: camera image microphone capabilities native device permissions media
ms.topic: conceptual
ms.author: lajanuar
---

> # Media capabilities in Teams

Reviewer 2: Should the title be integrate Media capabilities in Teams or Device capabilities. This needs more clarity. In the overview page we need to clearly differentiate between the two terms.

* Use native **microphone** to allow users to **record audio** (record 10 minutes of conversation) from the device.
* Use native **camera control** to allow users to **capture and attach images** on the go.
* Use native **gallery support** to allow users to **select device images** as attachments.
* Use native **image viewer control** to **preview multiple images** at one time.
* Support **large image transfer** (from 1 MB to 50 MB) through the SDK bridge.
* Support **advanced image capabilities** allowing users to preview and edit images:
  * Scan document, whiteboard, and business cards  through the camera.
  * Crop and rotate the images.
  * Add text, ink, or freehand annotation to the image.

> [!IMPORTANT]
>*   The `selectMedia`, `getMedia`, and `viewImages` APIs must be invoked from multiple Teams surfaces such as task modules, tabs, and personal apps. For more details, see [Entry points for Teams apps](../extensibility-points.md).
>* `selectMedia` API has been extended to support mic and audio properties.

You must use the following set of APIs to enable your device's media capabilities:

| API      | Description   |
| --- | --- |

> | [**selectMedia**](/javascript/api/@microsoft/teams-js/media?view=msteams-client-js-latest&branch=master#selectMedia_MediaInputs___error__SdkError__attachments__Media_______void_&preserve-view=true) for image capturing in camera| This API allows users to **capture or select media from a native device** and return to the web-app. The users edit, crop, rotate, annotate, or draw over images before submission. In response to **selectMedia**, the web-app receives the media IDs of selected images and a thumbnail of the selected media.|

Reviewer: ```suggestion
| [**selectMedia**](/javascript/api/@microsoft/teams-js/media?view=msteams-client-js-latest&branch=master#selectMedia_MediaInputs___error__SdkError__attachments__Media_______void_&preserve-view=true) for image capturing in camera| This API allows users to **capture or select media from device camera** and return it to the web-app. The users can edit, crop, rotate, annotate, or draw over images before submission. In response to **selectMedia**, the web-app receives the media IDs of selected images and a thumbnail of the selected media.|
```

> | [**selectMedia**](/javascript/api/@microsoft/teams-js/media?view=msteams-client-js-latest&branch=master#selectMedia_MediaInputs___error__SdkError__attachments__Media_______void_&preserve-view=true) for audio recording in **Microphone**| This API also allows the users to record audio from a native device and return to the web-app. The users pause, re-record, and play recording preview before submission. In response to **selectMedia**, the web-app receives media IDs of the selected audio recording. Set the media type to four for audio. <br/> Use `maxDuration`, if you require to configure a duration in minutes for recording the conversation. The current duration for recording is 10 minutes, after which the recording terminates.  |

Reviewer: ```suggestion
| [**selectMedia**](/javascript/api/@microsoft/teams-js/media?view=msteams-client-js-latest&branch=master#selectMedia_MediaInputs___error__SdkError__attachments__Media_______void_&preserve-view=true) for audio recording in **Microphone**| Set the media type to 4 in **selectMedia** API for accessing mic capability. This API also allows users to record audio from device mic and return recorded clip to the web-app. The users can pause, re-record, and play recording preview before submission. In response to **selectMedia**, the web-app receives media IDs of the selected audio recording. <br/> Use `maxDuration`, if you require to configure a duration in minutes for recording the conversation. The current duration for recording is 10 minutes, after which the recording terminates.  |
```

> | [**getMedia**](/javascript/api/@microsoft/teams-js/_media?view=msteams-client-js-latest&branch=master#getMedia__error__SdkError__blob__Blob_____void_&preserve-view=true)| This API retrieves the media in chunks irrespective of  the size. These chunks are assembled and sent back to the web app as a file or blob. This API breaks an image into smaller chunks, to facilitate large image transfer. |

Reviewer: ```suggestion
| [**getMedia**](/javascript/api/@microsoft/teams-js/_media?view=msteams-client-js-latest&branch=master#getMedia__error__SdkError__blob__Blob_____void_&preserve-view=true)| This API retrieves the media captured by **selectMedia* API in chunks irrespective of  the media size. These chunks are assembled and sent back to the web app as a file or blob. Breaking of media into smaller chunks facilitates large file transfer. |
```

| [**viewImages**](/javascript/api/@microsoft/teams-js/media?view=msteams-client-js-latest#viewImages_ImageUri_____error___SdkError_____void_&preserve-view=true)| This API enables the user to view images in  full-screen mode as a scrollable list.|

**Web app experience for selectMedia API for image capability**
![device camera and image experience in Teams](../../assets/images/tabs/image-capability.png)

**Web app experience for selectMedia API for microphone capability**
![web app experience for microphone capability](../../assets/images/tabs/microphone-capability.png)

## Error handling

> The following table lists the error codes and the conditions under which they are generated:

Reviewer: ```suggestion
The following table lists the error codes and the conditions under which they are generated. You must ensure that you handle these errors appropriately in your Teams app.
```

* Use the native **microphone** to allow users to **record audio** (record 10 minutes of conversation) from the device.
* Use native **camera control** to allow users to **capture and attach images** on the go.
* Use native **gallery support** to allow users to **select device images** as attachments.
* Use native **image viewer control** to **preview multiple images** at one time.
* Support **large image transfer** (from 1 MB to 50 MB) through the SDK bridge.
* Support **advanced image capabilities** allowing users to preview and edit images:
  * Scan document, whiteboard, and business cards  through the camera.
  
> [!IMPORTANT]
>*   The `selectMedia`, `getMedia`, and `viewImages` APIs can be invoked from multiple Teams surfaces such as task modules, tabs, and personal apps. For more details, see [Entry points for Teams apps](../extensibility-points.md).
>* `selectMedia` API has been extended to support mic and audio properties.

You must use the following set of APIs to enable your device's media capabilities:

| API      | Description   |
| --- | --- |

> | [**selectMedia**](/javascript/api/@microsoft/teams-js/media?view=msteams-client-js-latest&branch=master#selectMedia_MediaInputs___error__SdkError__attachments__Media_______void_&preserve-view=true) (**Camera)**| This API allows users to **capture or select media from the device camera** and return it to the web-app. The users can edit, crop, rotate, annotate, or draw over images before submission. In response to **selectMedia**, the web-app receives the media IDs of selected images and a thumbnail of the selected media.|

Reviewer: ```suggestion
| [**selectMedia**](/javascript/api/@microsoft/teams-js/media?view=msteams-client-js-latest&branch=master#selectMedia_MediaInputs___error__SdkError__attachments__Media_______void_&preserve-view=true) (**Camera)**| This API allows users to **capture or select media from the device camera** and return it to the web-app. The users can edit, crop, rotate, annotate, or draw over images before submission. In response to **selectMedia**, the web-app receives the media IDs of selected images and a thumbnail of the selected media. This API can be further configured via the [ImageProps](https://review.docs.microsoft.com/en-us/javascript/api/@microsoft/teams-js/imageprops?view=msteams-client-js-latest&branch=master) configuration. |
```

> | [**selectMedia**](/javascript/api/@microsoft/teams-js/media?view=msteams-client-js-latest&branch=master#selectMedia_MediaInputs___error__SdkError__attachments__Media_______void_&preserve-view=true) (**Microphone**)| Set the media type to `four` in **selectMedia** API for accessing microphone  capability. This API also allows users to record audio from the device microphone and return recorded clips to the web-app. The users can pause, re-record, and play recording preview before submission. In response to **selectMedia**, the web-app receives media IDs of the selected audio recording. <br/> Use `maxDuration`, if you require to configure a duration in minutes for recording the conversation. The current duration for recording is 10 minutes, after which the recording terminates.  |

Reviewer: ```suggestion
| [**selectMedia**](/javascript/api/@microsoft/teams-js/media?view=msteams-client-js-latest&branch=master#selectMedia_MediaInputs___error__SdkError__attachments__Media_______void_&preserve-view=true) (**Microphone**)| Set the [mediaType](https://review.docs.microsoft.com/en-us/javascript/api/@microsoft/teams-js/mediatype?view=msteams-client-js-latest&branch=master) to `4` in **selectMedia** API for accessing microphone  capability. This API also allows users to record audio from the device microphone and return recorded clips to the web-app. The users can pause, re-record, and play recording preview before submission. In response to **selectMedia**, the web-app receives media IDs of the selected audio recording. <br/> Use `maxDuration`, if you require to configure a duration in minutes for recording the conversation. The current duration for recording is 10 minutes, after which the recording terminates.  |
```

```

> [!NOTE]

> > The _Request Permissions_ prompt is automatically displayed when a relevant Teams API is initiated. For more information, see [Request device permissions](native-device-permissions.md).

Reviewer 2: markdown tag is incorrect and shouldn't it be bold tag?

* Support **advanced image capabilities** allowing users to preview and edit images:
  * Scan document, whiteboard, and business cards  through the camera.
  
> [!IMPORTANT]
>*   The `selectMedia`, `getMedia`, and `viewImages` APIs can be invoked from multiple Teams surfaces such as task modules, tabs, and personal apps. For more details, see [Entry points for Teams apps](../extensibility-points.md).
>* `selectMedia` API has been extended to support mic and audio properties.

You must use the following set of APIs to enable your device's media capabilities:

| API      | Description   |
| --- | --- |
| [**selectMedia**](/javascript/api/@microsoft/teams-js/media?view=msteams-client-js-latest&branch=master#selectMedia_MediaInputs___error__SdkError__attachments__Media_______void_&preserve-view=true) for image capturing in camera| This API allows users to **capture or select media from the device camera** and return it to the web-app. The users can edit, crop, rotate, annotate, or draw over images before submission. In response to **selectMedia**, the web-app receives the media IDs of selected images and a thumbnail of the selected media.|

> | [**selectMedia**](/javascript/api/@microsoft/teams-js/media?view=msteams-client-js-latest&branch=master#selectMedia_MediaInputs___error__SdkError__attachments__Media_______void_&preserve-view=true) for audio recording in **Microphone**| Set the media type to `four` in **selectMedia** API for accessing microphone  capability. This API also allows users to record audio from the device microphone and return recorded clips to the web-app. The users can pause, re-record, and play recording preview before submission. In response to **selectMedia**, the web-app receives media IDs of the selected audio recording. <br/> Use `maxDuration`, if you require to configure a duration in minutes for recording the conversation. The current duration for recording is 10 minutes, after which the recording terminates.  |

Reviewer 2: Lets just have (Microphone) in the selectMedia, suggest we can remove the other text.

| [**getMedia**](/javascript/api/@microsoft/teams-js/_media?view=msteams-client-js-latest&branch=master#getMedia__error__SdkError__blob__Blob_____void_&preserve-view=true)| This API retrieves the media captured by **selectMedia** API in chunks, irrespective of the media size. These chunks are assembled and sent back to the web app as a file or blob. Breaking of media into smaller chunks facilitates large file transfer. |
| [**viewImages**](/javascript/api/@microsoft/teams-js/media?view=msteams-client-js-latest#viewImages_ImageUri_____error___SdkError_____void_&preserve-view=true)| This API enables the user to view images in  full-screen mode as a scrollable list.|

**Web app experience for selectMedia API for image capability**
![device camera and image experience in Teams](../../assets/images/tabs/image-capability.png)

**Web app experience for selectMedia API for microphone capability**
![web app experience for microphone capability](../../assets/images/tabs/microphone-capability.png)

## Error handling

> The following table lists the error codes and the conditions under which the errors are generated. You must ensure to handle these errors appropriately in your Teams app.

Reviewer 2: Since you are using following, please add a colon instead of a period at the end of the sentence.

|Error code |  Error name     | Condition|
| --------- | --------------- | -------- |
| **100** | NOT_SUPPORTED_ON_PLATFORM | API is not supported on the current platform.|
| **404** | FILE_NOT_FOUND | File specified is not found in the given location.|
| **500** | INTERNAL_ERROR | Internal error is encountered while performing the required operation.|
| **1000** | PERMISSION_DENIED |Permission is denied by the user.|
| **2000** |NETWORK_ERROR | Network issue.|
| **3000** | NO_HW_SUPPORT | Underlying hardware does not support the capability.|
| **4000**| INVALID_ARGUMENTS | One or more arguments are invalid.|
| **5000** | UNAUTHORIZED_USER_OPERATION | User is not authorized to complete this operation.|
| **6000** |INSUFFICIENT_RESOURCES | Operation could not be completed due to insufficient resources.|
|**7000** | THROTTLE | Platform throttled the request as the API was invoked frequently.|
|  **8000** | USER_ABORT |User aborts the operation.|
| **9000**| OLD_PLATFORM | Platform code is outdated and does not implement this API.|
| **10000**| SIZE_EXCEEDED |  Return value is too big and has exceeded the platform size boundaries.|

## Code snippets

> **Calling `selectMedia` API** for capturing images using camera

Reviewer 2: Global: Please add a colon instead period.

});
```

> **Calling `getMedia` API** for capturing images using camera

Reviewer: ```suggestion
**Calling `getMedia` API** to retrieve large media in chunks
```

});
```

> **Calling `viewImages` API by ID**

Reviewer: ```suggestion
**Calling `viewImages` API by ID returned by `selectMedia` API**
```

```javascript

> view images by id:

Reviewer 3: First two lines should be comments. Could you please add // in front of first two lines in this snippet?

}
```

**Calling `viewImages` API by URL**:

```javascript

> View Images by URL:

Reviewer 3: Could you please add // in front of first line in this snippet? This is a comment.

Author: @Wajeed-msft , please approve the PR
